source/common/event/dispatcher_impl.cc

```diff
@@ -155,28 +153,46 @@ DispatcherImpl::createClientConnection(Network::Address::InstanceConstSharedPtr
                                                          std::move(transport_socket), options);
 }
 
+// @param dns_resolution_config always contains data since even DnsResolutionConfig is not present
+// in the caller protobuf config (could be bootstrap_, cluster, or dns_cache_config),
+// the caller will program some other data in it, like use_tcp_for_dns_lookups or dns_resolvers.
+// @param dns_resolver_config could be empty.
 Network::DnsResolverSharedPtr DispatcherImpl::createDnsResolver(
-    const std::vector<Network::Address::InstanceConstSharedPtr>& resolvers,
-    const envoy::config::core::v3::DnsResolverOptions& dns_resolver_options) {
+    envoy::config::core::v3::DnsResolutionConfig& dns_resolution_config,
+    envoy::config::core::v3::TypedExtensionConfig& dns_resolver_config) {
+
   ASSERT(isThreadSafe());
-#ifdef __APPLE__
-  static bool use_apple_api_for_dns_lookups =
+
+  Network::DnsResolverFactory* dns_resolver_factory;
+
+  if (dns_resolver_config.typed_config().type_url().empty()) {
+    // If the typed dns resolver configuration is missing, synthetic a dns_resolver_config
+    // based on the run time use_apple_api_for_dns_lookups flag. Then using
+    // the dns_resolver_config to retrieve dns_resolver_factory.
+
+    static bool use_apple_api_for_dns_lookups =
       Runtime::runtimeFeatureEnabled("envoy.restart_features.use_apple_api_for_dns_lookups");
-  if (use_apple_api_for_dns_lookups) {
-    RELEASE_ASSERT(
-        resolvers.empty(),
-        "defining custom resolvers is not possible when using Apple APIs for DNS resolution. "
-        "Apple's API only allows overriding DNS resolvers via system settings. Delete resolvers "
-        "config or disable the envoy.restart_features.use_apple_api_for_dns_lookups runtime "
-        "feature.");
-    RELEASE_ASSERT(!dns_resolver_options.use_tcp_for_dns_lookups(),
-                   "using TCP for DNS lookups is not possible when using Apple APIs for DNS "
-                   "resolution. Apple' API only uses UDP for DNS resolution. Use UDP or disable "
-                   "the envoy.restart_features.use_apple_api_for_dns_lookups runtime feature.");
-    return std::make_shared<Network::AppleDnsResolverImpl>(*this, api_.rootScope());
+    if (use_apple_api_for_dns_lookups) {
+      // In Apple system, ignore the dns_resolution_config.
+      dns_resolver_config.mutable_typed_config()->PackFrom(Envoy::ProtobufWkt::Struct());
+    } else {
+      dns_resolver_config.mutable_typed_config()->PackFrom(dns_resolution_config);
```

Comment:
I don't think this is quite right. In general in Envoy, we give each extension its own configuration message that uniquely implies the extension. This might even be empty. See for example https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/on_demand/v3/on_demand.proto.

So, we probable have some `envoy.extensions.dns.cares.v3.CAresConfig` message for c-ares, same for Apple etc.

Reply:
+1, we should not reuse the existing message for it, since the namespace does not match what extension would normally use. I think the best approach is to make it a copy of the `envoy::config::core::v3::DnsResolutionConfig` and then deprecate the `envoy::config::core::v3::DnsResolutionConfig`

I would also move the logic of dealing with the backward compatibility into a separate helper function in `source/network/dns_resolvers/common/` Then this method will be a lot cleaner.

Reply:
Thanks for the comments. Yes, added new message for each extension.  Also a separate function is added to deal with crafting the typed_dns_resolver_config for backward compatibility.